Add basic ltc codegen to staging branch

wconstab opened the pull request:
This is now a working minimal codegen prototype. I plan to merge to ltc_staging mostly as-is, but I'll fix anything folks point out.

As for what's not complete, there are a few known things that I plan to address (or have others work on) in following PRs, and get to a clean state before we'd merge the final thing back to pytorch master.

- has only been tested on a couple ops, but has been tested with at least one structured and unstructured op
- missing support for some dtypes in the codegen (process_ir_nodes) - should be easy to incrementally complete this, or, rigorously complete it in a focused effort.
- need refactoring for TS lowering code - see TODOs referring to hacks there

There is also some bikeshedding.  If folks have a suggestion now i'll change it, or it can be done later
- ~~'codegen' is the current name for the section of YAML that gets full codegen as opposed to `supported` ops which get decls not impls codegenned.~~ I renamed it to `full_codegen` which is hopefully less confusing
- names of generated code files are a little wonky (esp. LazyLazyIr.h)
- ~~organization of code in the generator .py files- i did a little bit to clean this up but maybe folks prefer other ways of grouping the code~~ this might be good now, I did a refactor after chatting with @bdhirsh

wconstab committed:
Duplicate existing Register* codegen machinery for new section of 'codegen' ops

wconstab committed:
WIP generate LazyIR.h output
currently working on how to filter to generate for only the requested
functions, then attempt to define and use a template for IR class

wconstab committed:
Generate IR node classes with mostly complete ctor/fields
- left input tensors (values) in undecided state between old
  way of passing vector of inputs to base class, and possibly
  just making them fields of the class like the scalars

wconstab committed:
WIP generate native function impls

ezyang commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,138 @@
+# Generates {backend}_lazy_ir.h and .cpp
+#
+valueListT = BaseCppType('at', 'ValueList') # TODO, not sure this type exists
+valueT = BaseCppType('ir', 'Value')
```
A thought: in the old code structure in XLA, we generate functions which take Value/ValueList as arguments and then call into them. I'm not sure this is really necessary: couldn't we directly pass Tensors into the generated IR construction code, and just have the IR construction convert the Tensor into a Value? Then you wouldn't need `process_ir_types` below.

wconstab replied:
It could still be a convenience, to put the conversion logic into the base 'node' class so that we don't have to codegen the conversions.  But yea, I guess there is no way around generating the correct 'Value' type in some part of the generated code.

ezyang commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,138 @@
+  NodePtr Clone(OpList operands) const override;
+
+ private:
+  std::vector<c10::Optional<ir::Value>> inputs;
```
It's kind of weird that all the scalars are unboxed, but the inputs are stashed into a vector. Maybe easier to just have them unboxed too.

wconstab replied:
Yea, I looked into this a bit and it's because the base class is tracking usages of Values for liveness purposes.  But this doesn't mean they have to be in a vector, it just means they need special handling, which we could generate another way.

ezyang commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,138 @@
+    def gen(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        # for now, we just want one IR class decl and soon after also the method defs
+        # and we use the functional version not out/inplace.
+        func = f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
```
Wouldn't it make more sense to just filter out the inplace/out codegen?

ezyang replied:
(then you would assert that it is functional here)

wconstab replied:
Where would I do the filtering, in the __call__ before it calls gen()?  Or in gen_backend_stubs somehow?  I figured there was a more appropriate way I was missing.

ezyang replied:
you can literally do it here, by returning an empty list. Other examples of filtering:

```
@with_native_function_and_index
def gen_unstructured(f: NativeFunction, backend_index: BackendIndex) -> Optional[str]:
    sig = kernel_signature(f, backend_index)
    metadata = backend_index.get_kernel(f)
    if metadata is None:
        return None
    if "legacy::" in metadata.kernel:
        return None
    else:
        prefix = 'static' if backend_index.external else 'TORCH_API'
        return f"{prefix} {sig.decl(name=metadata.kernel)};"
```

ezyang commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,138 @@
+        node_ctor_args = ", ".join([f"{i.cpp_type()} {i.name}" for i in all_types])
+        scalar_initializers = ",\n        ".join([f"{t.name}_({t.name})" for t in scalar_types])
+        scalar_decls = ";\n  ".join([f"{t.cpp_type()} {t.name}_" for t in scalar_types])
+        scalar_hashes = ", ".join([f.name for f in scalar_types])
```
This is definitely one of those places where, if we just changed the hashing convention, the codegen here would be so much simpler

wconstab replied:
What would be your proposed convention?  Ironically this version you commented on looks fairly simple compared to the nonsense I cobbled together to make things complile.  Taking 'optionals' into account for hashing turns out to be annoying.  Should I just make it so the native 'Hash' function accepts optionals natively and just doesn't include a contribution from them if they are null?

ezyang replied:
I think, if I didn't care about codegen friendliness, I'd probably have each child node compute the *full* hash, and so you just literally pass every argument into a hash function (which is some variadic template thing that successively combines the hashes of each argument). Do need to contribute to hash when optional is null in case of `optional<Tensor>, optional<Tensor>` need to distinguish between `nullopt, t` case and `t, nullopt` case

wconstab replied:
good point. I am already doing a nullopt thing for Value hashes actually. 

But what about scalars?  I hadn't thought about needing to distinguish between say 
optional<float>, optional<bool> as nullopt, false vs 1.0, nullopt.  How do I provide a 'null' value to hash on for scalars?

ezyang replied:
we talked about this in chat; if you don't need cryptographic hash you can just pick some random number like -3333 (like in glibc) for nullopt; for cryptographic hash (no collisions) need to distinguish these cases

ezyang commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,138 @@
+        if isinstance(t, BaseType):
+            if t.name == BaseTy.Tensor:
+                if mutable and not local.use_const_ref_for_mutable_tensors():
+                    types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))
```
You probably don't actually want to pass valueT by mutable reference

ezyang commented on tools/codegen/dest/native_functions.py:
```diff
@@ -53,3 +53,49 @@ def compute_native_function_declaration(
+{sig.decl(name=metadata.kernel)} {{
+    {lazy_tensors}
+    return bridge::AtenFromLtcTensor(l_{first_tensor.name}.CreateFrom(
+        ir::MakeNode<ir::ops::{metadata.kernel}>({node_ctor_inputs})));
```
It probably would make more sense to stop using `ir::MakeNode` and just arrange for the constructor on the IR class to be codegen friendly

ezyang replied:
Absent that, this is a good use case for `tools.codegen.api.translate`. If you can define a signature that describes what the types of the arguments ir::MakeNode is looking for, you can use translate to automatically convert from the source arguments to the target arguments, handling reordering and conversions automatically

wconstab replied:
so, ir::MakeNode is really just a wrapper around making a shared pointer.  I think what's more onerous is the pattern that we use the first tensor's lazytensor.CreateFrom() method.  I'm wondering if we can get rid of this and just create the node and wrap it in a new LazyTensor.

ezyang commented on tools/codegen/dest/native_functions.py:
```diff
@@ -53,3 +53,49 @@ def compute_native_function_declaration(
+    all_types, value_types, scalar_types = process_ir_types(f.func)
+
+    lazy_tensors = "\n    ".join([f"LazyTensor l_{t.name} = bridge::GetLtcTensor({t.name});" for t in value_types])
+    node_ctor_inputs = ", ".join([f"l_{t.name}" for t in value_types] +
```
It doesn't seem there is any particular reason to give locals for these, since they immediately get passed into MakeNode and are never subsequently used. Saves you from having to explicitly `std::move` them in too.

wconstab replied:
yea, i agree.  i guess i was just wary of making one line of code that was too hard to read.

ezyang commented:
looks pretty good, and pleasantly short!

wconstab committed:
WIP generate TS lowerings

wconstab committed:
Make it compile (with a few regrettable hacks)

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,175 @@
+                types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+                value_types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+            elif str(t.elem) == 'int':
+                types.append(NamedCType(binds, BaseCType(intArrayT)))
```
@ezyang What type should I use here?  I used a std::vector<int64_t> for now, but I was wondering if I should use at::Array or some other PyTorch type.

ezyang replied:
Can't use std::array/at::Array because you don't know what the size of the vector is. std::vector is fine. Maybe also consider SmallVector/DimVector

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,138 @@
+class LazyIR:
+    backend_index: BackendIndex
+
+    target: Union[
```
Yea.. I actually wanted to use it, but didn't like the names of the enum values for my cases.  I did end up using it in the TS lowering code, but for some reason didn't end up doing it that way for IR.  

(IR still generates both the 'class' def, and the 'construction site' code inside the native function.)

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,138 @@
+    # TODO(whc) probably use selector instead of building a separate index for ops to codegen
+    # Selector object to determine which operators to generate
+    # registration code for.
+    selector: SelectiveBuilder
```
I still have to manage 2 lists though.  One list gets only the registrations and declarations generated, and the second list gets those plus the implementations (and, for TS backend, the lowerings).  

For now I made things work with a couple hacks and a second list in the yaml.  It occured to me at the time that maybe selector was the right tool to use for deciding which ops to generate for a given TARGET or something like that.  But I'd be open to suggestions on how to handle this cleanly.

(I agree that as currently coded, I could just delete selector, and if we land on another means of solving the above issue i'll remove it)

wconstab commented on tools/codegen/dest/native_functions.py:
```diff
@@ -53,3 +53,49 @@ def compute_native_function_declaration(
+
+
+@with_native_function_and_index
+def gen_unstructured_lazy_definition(f: NativeFunction, backend_index: BackendIndex) -> Optional[str]:
```
OK with me.

Note that by now, I've got code in 3 dest files - lazy_ir, ts_lowering, and native_function.  These correspond to the name of the template where their output goes.  

In this case, I'm adding a .cpp version of the same-named NativeFunctions file, so it is kind of a new LTC-specific file even though it is related to the old one.

wconstab commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -264,5 +280,49 @@ def make_file_manager(install_dir: str) -> FileManager:
+                'backend_namespace': 'torch_lazy_tensors',  # this is wrong
+                'native_function_definitions': 
+                list(concatMap(
+                    lambda f: dest.compute_lazy_native_function_definition(f, backend_indices[CODEGEN_MAGIC_NUMBER]),
```
Yea, i definitely won't ship the hacky magic number.  But I wanted to get suggestions on a better way.  

I think I could make parse_backend_yaml return a separate list, and maybe figure out how to pass that list to my generators instead of using the backend index directly. 

Does that suggestion apply equally to the case I currently have in yaml where there is a separate list of ops for the new codegen vs the existing 'backend' list for the old codegen?  Or is this suggestion mainly for the case where i made it a 'flag' for each op whether to opt-in to the more complete codegen?

wconstab committed:
Suport hashing optionals

wconstab committed:
Implement IR ToString

wconstab committed:
Cleanup

wconstab committed:
Make it run

wconstab committed:
Implement incomplete ToString for c10::Scalar

wconstab committed:
Try codegen addcmul/addcdiv; seems to pass tests

wconstab committed:
Clean up how codegen list and backend index are handled

wconstab committed:
Fix c10::Scalar hash

wconstab committed:
Fix backend namespace of generated nativefunction impl

wconstab committed:
Fix native_function impl generation for NativeFunctionGroups

wconstab committed:
Make str_join compile

wconstab committed:
WIP Generate Clone methods, haven't tested yet

wconstab committed:
Remove old impl of mean(), WIP for SmoothL1

wconstab committed:
Refactor lazy type utils for cleanup and enablement of separate kwargs

Decouple the old process_ir* function which did type converison at the same
time as separation of values/scalars.  Make a better 'type conversion' function
and switch to packaging up a new FunctionSchema so it is easier to manipulate
fields of types such as the kwarg ones downstream.

This will help for generating TS lowerings which need to separate args/kwargs

wconstab committed:
WIP support kwarguments for TS lowering

wconstab committed:
WIP using meta kernel for cos shape, unittest passes

wconstab committed:
WIP generate stubs for compute shape/dtype for non-meta, runs but meancast still failing dtype test

wconstab committed:
Clean up more LAPACK test failures

wconstab committed:
Fix dtype bug for mean; not sure this approach is safe

wconstab committed:
Comment ShapeDtype

wconstab committed:
Clean up SmoothL1Loss hack

wconstab committed:
Restore disabled LAPACK tests for now, address these in a separate PR

wconstab commented on tools/codegen/dest/native_functions.py:
```diff
@@ -53,3 +58,138 @@ def compute_native_function_declaration(
+        dtypeFn = DispatcherSignature.from_schema(FunctionSchema(f.func.name,
+                                                                 f.func.arguments,
+                                                                 returns=(Return(name=None,
+                                                                                 type=scalarTypeT,
```
@bdhirsh I had trouble here.  Intended to create new DispatcherSignature objects with updated Returntypes, and use those to generate the code below (the function signature declarations).

The problem was if i called shapeFn.decl() it complained that scalarTypeT was invalid as a return type.  I didn't really understand the assumptions/logic of ReturnType, so I just gave up and wrote the code below.  

If you think there is a quick or longer-term-worthwhile fix for my above problem, i'll leave this code in place; if not, I can delete these and just use the code below.

wconstab committed:
Cleanup

wconstab committed:
Refactor native_functions.py

bdhirsh commented on aten/src/ATen/templates/NativeFunctions.cpp:
```diff
@@ -0,0 +1,10 @@
+// ${generated_comment}
```
Hmmm.. maybe rename to `DispatchKeyNativeFunctions.cpp`, like the existing `DispatchKeyNativeFunctions.h` template?

bdhirsh commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:
```diff
@@ -174,12 +180,30 @@ class TSNodeLowering : public NodeLowering {
+    // Desirable to have a way to do some ops via codegen and others by hand, 
+    // at least for now
+    auto codegen_lowering = LowerToTSCodegen(function_, loctx(), node);
+    if (codegen_lowering.size() > 0){
```
Just wondering what your thoughts are - I can also imagine a version of this where each node class has its own (virtual) shape inference function, which would get rid of the giant switch statement

wconstab replied:
Yep, I think that might be a good way to go.  XLA currently has it that way and Alex separated it out.  I think i'll leave that change as a separate PR if you're OK with it, since it probably makes sense to look holistically at the design of node class  rather than making the change reflexively

wconstab replied:
I think the main issue with having 'Lower' as a function on Node base is that Lower would return a backend-specific object.  We could rethink its API so it returns a general LowerResult class that holds an opaque reference to a real backend lower result.

bdhirsh replied:
Yeah - I guess right now, the two plugins still have some pretty large differences: the API each plugin needs to implement the `NodeLowering` interface, and xla and ts use two totally different classes under the hood to do it: `XlaOpVector and TSOpVector`. I could imagine that it'd be useful to try to figure out what the common interface between those types is so more of the logic could be implemented directly in LazyTensor instead of getting pushed to the plugins. But I bet that'll take a fair amount of design effort, and it seems like we can kind of paper over that for now with more codegen (aka what this PR currently does).

bdhirsh replied:
Also, totally not in scope for this PR but something I'm just trying to understand better: I see that this PR still has a second giant switch statement (which is *not* codegen'd) at the `Infer` function here:  https://github.com/pytorch/pytorch/pull/64104/files#diff-64c166802d4c26f15e7fafd4efe9f2c19def21142d732db9aa2b7b300eaa84fdR79. I'm still trying to figure out - is the `Infer` function orthogonal to the codegen'd shape inference that you're generating in this PR? Or can any of it be re-used (or can the Infer function even be killed in favor of the shape inference you're doing). I have a feeling this has to do a least partially with the difference between xla vs. aten shape inference, but I'll need to stare at it some more.

wconstab replied:
The link above didn't seem to be to an Infer function but I think I know what you're talking about.  I didn't want to change too many things at once.  We'll clean up the design of shape inference more holistically.  

One reason I didn't do more now is that I didn't want to have to codegen all the ops yet.  We still have ops which have manual TS lowerings, and for those we also have the manual TS shape inference pathway.  Once we move to codegen we'll naturally delete that code, and it'll move us toward the clean design.

On the XLA side, it's not clear yet if they will totally adopt our meta-inferred shapes, or if they will use it only for hashing but use their own lowering code to compute 'real' dtypes for example, and use those for some other parts.

bdhirsh commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -15,11 +15,12 @@
 # Returns a Tuple of (backend_key, autograd_key, cpp_namespace, updated BackendIndex mapping)
 ParsedExternalYaml = namedtuple('ParsedExternalYaml', [
-    'backend_key', 'autograd_key', 'cpp_namespace', 'backend_indices'])
+    'backend_key', 'autograd_key', 'cpp_namespace', 'backend_indices', 'codegen'])
```
I forget if I already said this - eventually, the external codegen will have more customers than just LTC and XLA, so we probably don't want everyone calling into the same codegen API if it includes some LTC-specific stuff (the codegen'd IR). Maybe there should be a separate `gen_ltc_stubs.py`.

I think it's probably fine to wait until onboarding another backend to think about that though.

bdhirsh replied:
actually - if you're unconditionally generating IR nodes / TS lowerings in this file, then maybe this should be a separate file. Since pytorch/xla and LTC/TS will want to generate separate things.

wconstab replied:
ok.  no issue making that change.  i guess i'll just do it now, no reason to delay.  And I guess i'll name it `gen_lazy_tensor` since ltc is a little vague and stubs is less relevant with complete codegen involved

bdhirsh commented on tools/codegen/gen_backend_stubs.py:
```diff
@@ -264,5 +275,105 @@ def make_file_manager(install_dir: str) -> FileManager:
+                    "ATen/MetaFunctions.h",
+                ]],
+                'native_functions_include': '',
+                'backend_namespace': 'torch_lazy_tensors',  # this is wrong
```
what is this going to be eventually?

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+                                                                                 type=intArrayT,
+                                                                                 annotation=None),)),
+                                                  prefix='compute_shape_')
+        dtypeFn = DispatcherSignature.from_schema(FunctionSchema(f.func.name,
```
yeah, referring to your other comment (https://github.com/pytorch/pytorch/pull/64104#discussion_r708741985), I think the lower version is better than shapeFn/dtypeFn here.

A more heavy-weight suggestion would be to add a new `LazySignature` type in `types.py`, similar to `DispatcherSignature`, that knows how to print it's normal/shape/dtype signatures. I'm not sure if that's worth it though.

bdhirsh commented on tools/codegen/dest/ts_lowering.py:
```diff
@@ -0,0 +1,104 @@
+from typing import List, Optional, Union, Tuple
```
do you think this should be prefixed with `lazy_` just so its clear which files in `codegen/dest` are relevant to LTC?

bdhirsh commented on tools/codegen/dest/ts_lowering.py:
```diff
@@ -0,0 +1,104 @@
+    backend_index: BackendIndex
+
+    # Names of operators we want to codegen for, a subset of backend_index
+    codegen: List[OperatorName]
```
nit: instead of storing the list of codegen'd op names here, you could probably use it directly in `gen_backend_stubs.py` to filter which NativeFunctions get passed into this class. So the invariant is `TsLowering(f)` is only called if f is meant to be codegen'd

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+valueT = BaseCppType('ir', 'Value')
+intArrayT = BaseCppType('std', 'vector<int64_t>')
+
+def process_ir_type(typ: BaseTy, mutable: bool, binds: str) -> NamedCType:
```
nit: I think type of `typ` is `Type`, not `BaseTy` (BaseTy is just the base enum, doesn't include stuff like list/optional).

Running mypy locally would probably catch this (and I usually make sure mypy passes before bothering to run the codegen; it catches a bunch of my dumb mistakes).

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+    if isinstance(typ, BaseType):
+        if typ.name == BaseTy.Tensor:
+            return NamedCType(binds, BaseCType(valueT))
+        elif typ.name == BaseTy.Scalar:
```
This function is a no-op if `not type.is_tensor_like()`, right? Maybe that would simplify some of this logic (you don't need to deal with all of the non-tensor cases)

wconstab replied:
@ezyang actually steered me away from calling back into argumenttype_type and encouraged me to reimplement the whole logic here.  I ended up agreeing that this way is a bit clearer.

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+
+valueListT = BaseCppType('std', 'vector<ir::Value>')
+valueT = BaseCppType('ir', 'Value')
+intArrayT = BaseCppType('std', 'vector<int64_t>')
```
technically this is kinda redundant; I think you should be able to use `VectorCType(BaseCType(intT))` from `types.py`

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+from tools.codegen.api.translate import translate
+from tools.codegen.selective_build.selector import SelectiveBuilder
+
+valueListT = BaseCppType('std', 'vector<ir::Value>')
```
and just to follow the way that C++ containers are currently represented; in every place where you're currently using `BaseCType(valueListT)`, you should be able to use `VectorCType(BaseCType(valueT))`

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+    being Tensor-like, but assumes the type has already been transformed.
+    """
+    if isinstance(typ.type, BaseCType):
+        return typ.type.type.ns == valueT.ns and typ.type.type.name == valueT.name
```
does `return type.type.type == valueT` not work?

You'll also have to a case for `isinstance(typ.type, VectorCType)`

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+            new_args_fields[arg_field] = [
+                Argument(
+                    arg.name,
+                    process_ir_type(arg.type, arg.is_write, arg.name), arg.default, arg.annotation) for arg in getattr(func.arguments, arg_field)]
```
Hmm... transforming the `FunctionSchema` feels a little sketchy. I also see that `process_ir_type` returns a `NamedCType`, but the constructor for `Argument()` only takes in normal `Type`s?

It looks like the main purpose of this function is to substitute `Tensor` with `ir::Value` in a bunch of places. Instead of transforming the `FunctionSchema`, why can't you do that translation when you're generating each C++ string?

wconstab replied:
I had messier code before, and i refactored to manipulate FunctionSchema since it seemed like a nice way to modularize some of the transformations i wanted to do.

is your main objection that I abuse an existing class that has some semantic meaning, and if I make a new class for Lazy* it'll be cleaner?

Or do you also object to the modification at `FunctionSchema` level vs `Signature` level?  

I do more than s/Tensor-like/Value-like.  I also do stuff to make reference types into concrete types I can store in a class.

Also, I use the resulting modified schema in more places than just generating a decl or defn for a function.  For example I use the resulting modified argument types in places for IR node construction, and, I find it useful to have access to the modified arguments from the perspective of pre_self_positional, vs kwargs, etc - see how I use that in ts_lowering.py

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+        else:
+            node_ctor_scalars.append(t.name)
+
+    node_ctor_inputs = ",\n                              ".join(node_ctor_values + node_ctor_scalars)
```
i hate this but I've done this before too 😛

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -0,0 +1,387 @@
+    return node_ctor_inputs
+
+
+def ir_node_name(func: FunctionSchema):
```
Okay, I know we talked about this a little before - now I am wondering if there should just be an `api/lazy.py` file that contains most of these utility function mappings from `FunctionSchema` objects to various C++ names/strings/types.

If you look at `api/dispatcher.py`, it has several functions like:
`def name(func: FunctionSchema) -> str`
`def argument_type(a: Argument, *, binds: ArgName) -> NamedCType`
`def returns_type(rs: Sequence[Return]) -> CType`

each of which do some dispatcher-specific mapping, but fall back to the same api's in `api.cpp.py` in most cases. That sounds pretty similar to what most of these utility functions do. @ezyang

bdhirsh commented:
I think I'm done a round of review. Looks good!

wconstab committed:
Rename NativeFunctions template to match existing DispatchKey hdr

wconstab committed:
Factor out new gen_lazy_tensor.py

wconstab committed:
Make gen_ts_lowerings an optional feature of gen_lazy_tensor.py

wconstab committed:
Update .gitignore

wconstab committed:
Simplify compute_ function decls

wconstab committed:
Rename TSLowering

wconstab committed:
Fix process_ir_type typ mypy

desertfire commented on lazy_tensor_core/lazy_tensor_core/csrc/tensor_ops.cpp:
```diff
@@ -1,7 +1,9 @@
+#include "lazy_tensor_core/csrc/aten_ltc_bridge.h"
 #include "lazy_tensor_core/csrc/helpers.h"
 #include "lazy_tensor_core/csrc/ir.h"
+#include "lazy_tensor_core/csrc/ts_backend/LazyLazyIr.h"
```
Is this .h naming only temporary?

wconstab replied:
it got generated from the pattern {backend_key}LazyIr.h. 

It happens that our 2 backends have the key 'xla' and 'lazy' for now. 

I'm not sure what we're gonna do for naming of backends later on, so I didn't think about this too much yet.

perhaps for now, I can just remove the second 'Lazy' so we get 'LazyIr' and 'XlaIr', but, i am not sure if that is actually clearer or might be confusing to XLA folks who actually have their own IR

wconstab committed:
Some mypy influenced refactoring of types

wconstab committed:
refactor lazy_ir helpers into api/lazy.py

wconstab committed:
Refactor lazy utils to LazyIrSchema class

wconstab committed:
Fix more mypy

wconstab committed:
Stop abusing Schema and Argument classes, and make mypy pass

wconstab committed:
Clean up how codegen filtering works

wconstab committed:
fix mypy again

wconstab committed:
Whitespace

wconstab committed:
Merge branch 'lazy_tensor_staging' into wconstab/ltc_codegen

wconstab committed:
Merge branch 'lazy_tensor_staging' into wconstab/ltc_codegen

wconstab committed:
Whitespace

wconstab committed:
Fix newlines

Krovatkin commented on tools/codegen/api/lazy.py:
```diff
@@ -0,0 +1,175 @@
+        self.returns = func.returns
+
+    @property
+    def node_name(self) -> str:
```
we could probably add this property to an yaml for cases where we would like to have custom mappings such as `aten::expand`

wconstab replied:
oh, this isn't really for custom names.  its just for capitalized names for following c++ naming 
e.g. aten::mean op gets codegenned IR class named 
class Mean 

and underscores should get split out and each word-part Capitalized...

so i just put it here as a helper for parts of the codegen that use it

Krovatkin reviewed (commented):
Some initial feedback. I wanna go more in depth on this PR

wconstab commented on lazy_tensor_core/lazy_tensors/str_join.h:
```diff
@@ -10,6 +12,7 @@
+
 template <typename Range>
 std::string StrJoin(const Range& range, c10::string_view separator) {
   auto b = std::begin(range);
```
ah, excellent. one long-term TODO for the whole project is to start migrating to our native utils instead of the XLA copied ones, so we should be looking for these along the way.  I'll make the change.

wconstab commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -0,0 +1,386 @@
+        yaml_values = yaml.load(f, Loader=YamlLoader)
+    assert isinstance(yaml_values, dict)
+
+    valid_keys = ['backend', 'cpp_namespace', 'extra_headers', 'supported', 'autograd', 'codegen']
```
yup.  right now the names are dumb (need bikeshed), but 
supported:  it codegens the dispatcher registrations and nativefunction headers and you provide the impls
codegen: it codegens the above plus nativefunc impls, IR nodes, and optionally, TS lowerings

wconstab replied:
yes.

Krovatkin replied:
I think a comment(s) what `supported`, `codegen` mean would help more than a good name here :)

wconstab commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -0,0 +1,386 @@
+    cpp_namespace = yaml_values.pop('cpp_namespace', None)
+    assert cpp_namespace is not None, 'You must provide a value for "cpp_namespace"'
+
+    supported = yaml_values.pop('supported', [])
```
yea, we build a 'backend index' out of all the supported ops.  then we just track a separate list of the ops that get extra codegen.

wconstab commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -0,0 +1,386 @@
+            op_name = OperatorName.parse(op)
+            assert op_name in native_functions_map, f"Found an invalid operator name: {op_name}"
+            # See Note [External Backends Follow Dispatcher API]
+            kernel_name = dispatcher.name(native_functions_map[op_name].func)
```
i don't know what you mean

wconstab commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -0,0 +1,386 @@
+
+def parse_backend_yaml(
+        backend_yaml_path: str,
+        grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
```
hmm. i actually copied this code from `gen_backend_stubs.py` and didn't notice that; 

I am not sure if its a big deal that we traverse the `grouped_native_functions` twice, but perhaps more worthwhile is whether we could share more of this copied (and somewhat modified) code with gen_backend_stubs

bdhirsh replied:
Yeah, it would be nice if we could share most of the code this and gen_backend_stubs.py, so they don't diverge too much over time (and at least right now, almost everything used in `gen_backend_stubs.py` is also used here.

bdhirsh replied:
One option: You could just directly call the `parse_backend_yaml` function from `gen_backend_stubs.py`, then write a separate (much smaller) function that parses out the `codegen` list. That would require reading the yaml file twice though (hopefully that wouldn't noticeable slow down the build), and you'd also have to update `parse_backend_yaml` to not error out when it sees an unexpected yaml entry (`codegen`).

And you could even wrap all of the code that generates `Register{Lazy}.cpp` into a function and re-use it here too.

wconstab commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -0,0 +1,386 @@
+            })
+
+        # and generate headers for shape/dtype funcs for non-meta kernels
+        for dispatch_key in [backend_dispatch_key]:  # , autograd_dispatch_key
```
hmm, i didn't pay attention to the autograd kernels.  will this be relevant? i expect what will happen is that ops with autograd kernels don't get IR nodes anyway, and its ok that we continue to rely on the old form of codegen-stubs for those functions.

iiuc the autograd kernels are composite kernels that store some state on the backward tape.  then later, lazytensor will trace through those stored ops and build IR for them.  so i don't think we ever want to codegen an IR corresponding to the autograd kernel itself.  (not to be confused with a 'real' kernel that is used by autograd during backward, which we _do_ want IR for)

wconstab replied:
woah, i just realized all my code has a for loop over `[backend_dispatch_key, autograd_dispatch_key]`.

This should probably be deleted per my comment above.  

I think in my testing, the 'filtering' for codegen ops blocks the `autograd_dispatch_key` from ever being used with the lazy extra codegen.

Krovatkin replied:
> nevermind, I had already commented out the autograd key.

hehe this is pretty much the outcome I was hoping for here 😸

wconstab commented on lazy_tensor_core/ts_native_functions.yaml:
```diff
@@ -1,10 +1,14 @@
 backend: Lazy
 cpp_namespace: torch_lazy_tensors
+codegen:
```
i turned on the codegen for a small set of ops in this PR.  once we agree on the structure of the codegen infra I intend to
(1) land as-is to staging branch
(2) incrementally enable more ops (deleting the handcoded version of the op, verifying tests) and fix codegen issues as they come up for handling more ops
(3) finally merge the codegen to master

Krovatkin replied:
@wconstab my bad! I should've read the PR description more carefully!

wconstab committed:
Cleanup

wconstab committed:
flake8

wconstab committed:
Update gitignore

wconstab committed:
Rename codegen->full_codegen as YAML section

alanwaketan commented on aten/src/ATen/templates/LazyTsLowering.cpp:
```diff
@@ -0,0 +1,53 @@
+namespace torch_lazy_tensors {
+namespace compiler {
+
+// I copied these for now but they really should be moved.
```
Nit: Good to add a TODO here such that it's easier to spot in the future.

alanwaketan commented on aten/src/ATen/templates/LazyTsLowering.cpp:
```diff
@@ -0,0 +1,53 @@
+        const auto tuple_call_result = sv->asTuple({}, *function);
+        TSOpVector tuple_result;
+        for (const auto& tuple_component : tuple_call_result) {
+        auto tuple_component_sv =
```
Nit: Wrong indention.

alanwaketan commented on aten/src/ATen/templates/LazyTsLowering.cpp:
```diff
@@ -0,0 +1,53 @@
+    const ir::Node* node,
+    const std::vector<torch::jit::NamedValue>& arguments,
+    const std::vector<torch::jit::NamedValue>& kwarguments = {}) {
+return LowerBuiltin(function, node->op().op, arguments, kwarguments);
```
Nit: Wrong indention.

alanwaketan commented on aten/src/ATen/templates/LazyTsLowering.cpp:
```diff
@@ -0,0 +1,53 @@
+}
+
+TSOpVector LowerBuiltin(
+    std::shared_ptr<torch::jit::GraphFunction> function,
```
Can we make this function const reference?

alanwaketan replied:
I mean `const std::shared_ptr<torch::jit::GraphFunction>& function`.

alanwaketan commented on aten/src/ATen/templates/LazyTsLowering.cpp:
```diff
@@ -0,0 +1,53 @@
+
+${lowering_definitions}
+
+TSOpVector LowerToTSCodegen(std::shared_ptr<torch::jit::GraphFunction> function,
```
Ditto.

alanwaketan commented on aten/src/ATen/templates/LazyTsLowering.cpp:
```diff
@@ -0,0 +1,53 @@
+                            const ir::Node* node) {
+    switch (node->op().op){
+${lowering_dispatches}
+default:
```
Nit: wrong indentation.

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/tensor_ops.cpp:
```diff
@@ -122,8 +124,18 @@ LazyTensor SmoothL1Loss(const LazyTensor& input, const LazyTensor& target,
     case ReductionMode::kMean:
-      return LazyTensor::mean(elementwise_loss, all_dimensions, false,
-                              broadcasted_input.dtype());
+      //TODO(whc) SmoothL1Loss is not implemented by lazy TS backend,
```
Nit: wong indention.

wconstab replied:
this whole file seems wrong, i think my change fits in with the surroundings right?  prefer not to change the whole file in the same PR as other meaningful changes

alanwaketan replied:
Oops, I mean `//TODO(whc)` => `// TODO(whc)`.

wconstab replied:
ahh 👍

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/LazyShapeDtype.cpp:
```diff
@@ -0,0 +1,31 @@
+namespace ir {
+namespace ops {
+
+std::vector<int64_t> compute_shape_mean(const at::Tensor & self, c10::optional<at::ScalarType> dtype) {
```
Nit: Since the code is handwritten, maybe it's better to have `const at::Tensor& self` instead.

wconstab replied:
i made the change, but i'm not sure it's useful unless we also fix the generator so it generates matching headers.  The reason is that it's nice to just copy-paste the generated declaration and fill in the implementation.   i'll add a todo here to consider updating the generator correspondingly

wconstab replied:
actually I don't know if it's worth fixing the generator for this case. might be ok to just have the inconsistency in this file.

Krovatkin replied:
it will probably become more relevant when we start running  `clang-format` /`clang-tidy` and would want to avoid warnings, noise in diff due to formatting changes etc.

alanwaketan replied:
For the code-gen header, it probably doesn't matter.

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/LazyShapeDtype.cpp:
```diff
@@ -0,0 +1,31 @@
+std::vector<int64_t> compute_shape_mean(const at::Tensor & self, c10::optional<at::ScalarType> dtype) {
+    return std::vector<int64_t>({1});
+}
+c10::ScalarType compute_dtype_mean(const at::Tensor & self, c10::optional<at::ScalarType> dtype) {
```
Ditto.

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:
```diff
@@ -46,10 +46,16 @@
 namespace compiler {
 
+// Forward decl of method that is codegenned and has no header
+TSOpVector LowerToTSCodegen(std::shared_ptr<torch::jit::GraphFunction> function,
```
Not sure how this is linked to LazyTsLowering.cpp.

wconstab replied:
This is the generated function that has the switch statement that calls into the specific lowering function per IR class.

alanwaketan replied:
I understand that part. Just not quite exactly sure how this file is linked (linker) to that particular implementation.

wconstab replied:
oh. nothing special, it's just how the linker works, I think! doing this forward decl is exactly the same as putting this line into a header file and including the header file here.  it causes the linker to declare an undefined `LowerToTSCodegen` in this compilation unit, and then later when the generated code .o is linked together with ts_node_lowerin.o into the same .so lib, this becomes resolved.

alanwaketan replied:
Sure, if there is only one definition. What happens if there are two?

wconstab replied:
I don't think there is any difference.  You can still mistakenly define foo in 2 compilation units, even if you use a header.  Even without the header, if you define it twice in the same link unit it'll complain in the same way as it complains if there were a header.  (Multiple definition error)

alanwaketan replied:
This intrigues me because I have never done it or seem people forward declare a function without a header. And therefore wonder what could be the possible downsides.

wconstab replied:
oh- to be clear, I'm not defending that a forward decl is a nice practice for our codebase.  I just did it out of laziness.  I think there's a >50% chance we completely redo the way the lowering code is architected.  One option is move all the lowering funcs as class methods on the node.  So, I didn't want to waste time making a new header file, although, that is silly because it is very easy to make a new file too.

alanwaketan reviewed (commented):
This is great! Thanks for making this code-gen infra real.

wconstab committed:
Refactor to share parse and gen code with gen_backend_stubs

wconstab committed:
Fix PR feedback

bdhirsh reviewed (approved):
It looks like there's plenty of future work required, but this looks good as a first version of the codegen!

Krovatkin commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/LazyShapeDtype.cpp:
```diff
@@ -0,0 +1,31 @@
+namespace ops {
+
+std::vector<int64_t> compute_shape_mean(const at::Tensor& self, c10::optional<at::ScalarType> dtype) {
+    return std::vector<int64_t>({1});
```
this will be updated in future to support cases where we don't reduce the whole tensor??

wconstab replied:
i think that case is disambiguated by having a different set of function arguments (maybe even name?). -- so we'd generate another `compute_shape_mean` function which included the `dim` argument.

note: for some reason `mean` by itself has no meta kernel yet, but i think mean.dim does have a meta kernel.  so for mean.dim, we'd never have to provide this function, just rely on meta.

Krovatkin reviewed (approved):
:shipit:

wconstab committed:
Fix long strings

wconstab committed:
Fix more formatting

wconstab committed:
Fix flake8
